src/lib.mo

```diff
@@ -0,0 +1,327 @@
+    };
+
+    func removeValue(id : Nat) : () = values.put(id, null);
+
+    /// Dump all current stats to array
+    public func dump() : [(Text, Nat)] {
+      let result = Vector.Vector<(Text, Nat)>();
+      for (v in values.vals()) {
+        switch (v) {
+          case (?value) Vector.addFromIter(result, Iter.fromArray(value.dump()));
+          case (null) {};
+        };
+      };
+      Vector.toArray(result);
+    };
+
+    func renderSingle(name : Text, value : Text, timestamp : Text) : Text = name # " " # value # " " # timestamp # "\n";
+
+    /// Render all current stats to prometheus format
+    public func renderExposition() : Text {
+      let timestamp = Int.toText(now() / 1000000);
+      var res = "";
+      for ((name, value) in dump().vals()) {
+        res #= renderSingle(name, Nat.toText(value), timestamp);
+      };
+      res;
+    };
+
+    /// Dump all values, marked as stable, to stable data structure
+    public func share() : StableData {
+      var res : StableData = null;
+      for (value in values.vals()) {
+        switch (value) {
+          case (?v) switch (v.share()) {
+            case (?data) {
+              res := AssocList.replace(res, v.prefix, Text.equal, ?data).0;
+            };
+            case (_) {};
+          };
+          case (null) {};
+        };
+      };
+      res;
+    };
+
+    /// Patch all values with stable data
+    public func unshare(data : StableData) : () {
+      for (value in values.vals()) {
+        switch (value) {
+          case (?v) switch (AssocList.find(data, v.prefix, Text.equal)) {
+            case (?data) v.unshare(data);
+            case (_) {};
+          };
+          case (_) {};
+        };
+      };
+    };
+
+  };
+
+  class PullValue(prefix_ : Text, pull : () -> Nat) {
+    public let prefix = prefix_;
+
+    public func dump() : [(Text, Nat)] = [(prefix # "{}", pull())];
+
+    public func share() : ?StableDataItem = null;
+    public func unshare(data : StableDataItem) = ();
+  };
+
+  class CounterValue(prefix_ : Text, isStable : Bool) {
+    public let prefix = prefix_;
+
+    public var value = 0;
+
+    public func add(n : Nat) { value += n };
+    public func set(n : Nat) { value := n };
+
+    public func dump() : [(Text, Nat)] = [(prefix # "{}", value)];
+
+    public func share() : ?StableDataItem {
+      if (not isStable) return null;
+      ? #counter(value);
+    };
+    public func unshare(data : StableDataItem) = switch (data, isStable) {
+      case (#counter x, true) value := x;
+      case (_) {};
+    };
+  };
+
+  class GaugeValue(prefix_ : Text, buckets_ : ?[Nat], now : () -> Int, watermarkResetIntervalSeconds : Nat, isStable : Bool) {
+    public let prefix = prefix_;
+    public let buckets : ?[Nat] = switch (buckets_) {
+      case (null) null;
+      case (?b) ?Array.sort(b, Nat.compare);
+    };
+
+    class WatermarkTracker<T>(default : T, condition : (old : T, new : T) -> Bool, resetIntervalSeconds : Nat) {
```

Comment:
Since we have a class for this maybe we should put the bucketing code also in a class like `BucketTracker`? That would also remove the `null` case (0 buckets) from the internal code of that class.